# Multiple Regression

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("carfulldata.csv")

In [3]:
df.head()

Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105

In [4]:
X = df[["Weight", "Volume"]]
y = df["CO2"]

In [5]:
X.head(3)

Weight  Volume
0     790    1000
1    1160    1200
2     929    1000

In [6]:
y.head(3)

0    99
1    95
2    95
Name: CO2, dtype: int64

<b>Tip:</b> It is common to name the list of independent values with a upper case X, and the list of dependent values with a lower case y.


In [7]:
from sklearn import linear_model

In [8]:
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression()

In [9]:
predictionco2 = regr.predict([[4200, 2000]])

D:\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [60]:
print(predictionco2)

[127.01921288]


# Full Code

In [10]:
import pandas as pd
from sklearn import linear_model

In [11]:
df = pd.read_csv("carfulldata.csv")
X = df[["Weight", "Volume"]]
y = df["CO2"]

In [12]:
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression()

In [14]:
predictco2 = regr.predict([["3000", "1300"]])
print(predictco2)

[112.49439589]


D:\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
D:\Anaconda3\lib\site-packages\sklearn\base.py:566: FutureWarning: Arrays of bytes/strings is being converted to decimal numbers if dtype='numeric'. This behavior is deprecated in 0.24 and will be removed in 1.1 (renaming of 0.26). Please convert your data to numeric values explicitly instead.
  X = check_array(X, **check_params)


So, We have predicted that a car with 1.59 liter engine, and a weight of 2150 kg, will release approximately 108 grams of CO2 for every kilometer it drives.

# Coefficient

The coefficient is a factor that describes the relationship with an unknown variable.

Example: if x is a variable, then 2x is x two times. x is the unknown variable, and the number 2 is the coefficient.

In this case, we can ask for the coefficient value of weight against CO2, and for volume against CO2. The answer(s) we get tells us what would happen if we increase, or decrease, one of the independent values.

In [15]:
import pandas as pd
from sklearn import linear_model

df = pd.read_csv("carfulldata.csv")
df.head(4)

X = df[["Weight", "Volume"]]
y = df["CO2"]

In [16]:
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression()

In [17]:
# Coefficient as coef_
print(regr.coef_)

[0.00755095 0.00780526]


# Result Explained

The result array represents the coefficient values of weight and volume.

Weight: 0.00755095
Volume: 0.00780526

These values tell us that if the weight increase by 1kg, the CO2 emission increases by 0.00755095g.

And if the engine size (Volume) increases by 1 cm3, the CO2 emission increases by 0.00780526 g.

In [18]:
import pandas as pd
from sklearn import linear_model

In [19]:
df = pd.read_csv("carfulldata.csv")

X = df[["Weight", "Volume"]]
y = df["CO2"]

In [20]:
regr = linear_model.LinearRegression()
regr.fit(X, y)

LinearRegression()

In [21]:
prediction = regr.predict([[4000, 1300]])

D:\Anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [49]:
print(prediction)

[114.75968007]


So in this code we've predicted that a car with 1.3 litre engine, and a weight of 3300 kilogram, will release approximately 115 grams of co2 for every kilometer it drives.

which shows that the coefficient of 0.00755095 is correct:

106.07609071 + (1000 * 0.00755095) = 115.96
